asserts/preseed.go

```diff
@@ -140,10 +147,84 @@ func checkPreseedSnap(snap map[string]interface{}) (*PreseedSnap, error) {
 		return nil, fmt.Errorf("snap id is required when revision is set")
 	}
 
+	var components []PreseedComponent
+	if comps, ok := snap["components"]; ok {
+		components, err = checkPreseedComponents(comps, snapRevision)
+		if err != nil {
+			return nil, err
+		}
+	}
+
 	return &PreseedSnap{
+		Name:       name,
+		SnapID:     snapID,
+		Revision:   snapRevision,
+		Components: components,
+	}, nil
+}
+
+func checkPreseedComponents(comps interface{}, snapRevision int) ([]PreseedComponent, error) {
+	const wrongHeaderType = `"components" header must be a list of maps`
+
+	entries, ok := comps.([]interface{})
+	if !ok {
+		return nil, errors.New(wrongHeaderType)
+	}
+
+	seen := make(map[string]bool)
+	components := make([]PreseedComponent, 0, len(entries))
+	for _, entry := range entries {
+		comp, ok := entry.(map[string]interface{})
+		if !ok {
+			return nil, errors.New(wrongHeaderType)
+		}
+
+		preseedComp, err := checkPreseedComponent(comp, snapRevision)
+		if err != nil {
+			return nil, err
+		}
+
+		if _, ok := seen[preseedComp.Name]; ok {
+			return nil, fmt.Errorf("cannot list the same component %q multiple times", preseedComp.Name)
+		}
+
+		seen[preseedComp.Name] = true
+		components = append(components, preseedComp)
+	}
+
+	return components, nil
+}
+
+func checkPreseedComponent(comp map[string]interface{}, snapRevision int) (PreseedComponent, error) {
+	name, err := checkNotEmptyStringWhat(comp, "name", "of component")
+	if err != nil {
+		return PreseedComponent{}, err
+	}
+
+	if err := naming.ValidateSnap(name); err != nil {
+		return PreseedComponent{}, err
+	}
+	what := fmt.Sprintf("of component %q", name)
+
+	var revision int
+	if _, ok := comp["revision"]; ok {
+		revision, err = checkSnapRevisionWhat(comp, "revision", what)
+		if err != nil {
+			return PreseedComponent{}, err
```

Comment:
and this error path are not tested

Reply:
Added in 1ab195d1ea43aab36d1cbe33e6a8936f07a4d44c.